### PR TITLE
feat(flutter): date picker improvements

### PR DESCRIPTION
**Pull request by richardshiue**
Screenshot 2024-10-19 at 5 37 53 PM

User-facing changes:

- enter date and time using keyboard, support parsing in current locale
- improve look of the date picker header, no longer look squished on Linux and WIndows platforms
- i18n for database date cells
- focus traversal and submit on leave

Developer-facing changes:

- clean up parameter and callbacks to `AppFlowyDatePicker` and `MobileAppFlowyDatePicker` widgets.
- internal states for these widgets are better encapsulated and managed within the widget, ensuring same behavior across different places where the widget is used. No longer worry about `focusedDay`, `selectedDay`, `startDay`, `endDay` parameters, but just pass in `dateTime`, `endDateTime`, `includeTime`, `isRange` and desired onChanged callbacks.

resolves https://github.com/AppFlowy-IO/AppFlowy/issues/4617
resolves https://github.com/AppFlowy-IO/AppFlowy/issues/4309
resolves https://github.com/AppFlowy-IO/AppFlowy/issues/3496
resolves https://github.com/AppFlowy-IO/AppFlowy/issues/3027
resolves https://github.com/AppFlowy-IO/AppFlowy/issues/2441
resolves https://github.com/AppFlowy-IO/AppFlowy/issues/1887

### Feature Preview

---

#### PR Checklist

- [x] My code adheres to [AppFlowy's Conventions](https://docs.appflowy.io/docs/documentation/software-contributions/conventions)
- [x] I've listed at least one issue that this PR fixes in the description above.
- [x] I've added a test(s) to validate changes in this PR, or this PR only contains semantic changes.
- [x] All existing tests are passing.

### CODE REVIEW

File: frontend/appflowy_flutter/lib/plugins/database/application/cell/bloc/date_cell_editor_bloc.dart [hunk 1]
```diff
@@ -115,241 +112,129 @@ class DateCellEditorBloc
...
-      endDate: newEndDate,
-      endTime: newEndTime,
-      includeTime: includeTime ?? state.includeTime,
-      isRange: isRange ?? state.isRange,
-      reminderId: reminderId ?? state.reminderId,
+      date: dateTime,
+      endDate: endDateTime,
+      isRange: isRange,
     );
+    result.fold((_) {}, (err) => Log.error(err));
```
Comment (LucasXu0) [inline frontend/appflowy_flutter/lib/plugins/database/application/cell/bloc/date_cell_editor_bloc.dart:221]: If we don't need to handle the success case, we can use `result.onFailure` instead.

File: frontend/appflowy_flutter/lib/workspace/presentation/widgets/date_picker/appflowy_date_picker.dart [hunk 2]
```diff
@@ -1,181 +1,139 @@
...
-  void didUpdateWidget(covariant AppFlowyDatePicker oldWidget) {
-    _selectedDay = oldWidget.selectedDay != widget.selectedDay
-        ? widget.selectedDay
-        : _selectedDay;
-    _selectedReminderOption =
-        oldWidget.selectedReminderOption != widget.selectedReminderOption
-            ? widget.selectedReminderOption
-            : _selectedReminderOption;
+  void didUpdateWidget(covariant oldWidget) {
+    setState(() {
```
Comment (LucasXu0) [inline frontend/appflowy_flutter/lib/workspace/presentation/widgets/date_picker/appflowy_date_picker.dart:110]: > The framework always calls build after calling didUpdateWidget, which means any calls to setState in didUpdateWidget are redundant.

The setState should be reundant here.

File: frontend/appflowy_flutter/lib/workspace/presentation/widgets/date_picker/mobile_appflowy_date_picker.dart [hunk 3]
```diff
@@ -18,146 +18,317 @@ import 'package:go_router/go_router.dart';
...
+              Navigator.of(context).pop();
+            },
+          ),
         ],
         const _Divider(),
       ],
     );
   }
+
+  void onDateSelectedFromDatePicker(
```
Comment (LucasXu0) [inline frontend/appflowy_flutter/lib/workspace/presentation/widgets/date_picker/mobile_appflowy_date_picker.dart:190]: These two date pickers share similar logic. Should these functions, like `onDateSelectedFromDatePicker`, be composed together or managed using a bloc?

File: frontend/appflowy_flutter/lib/plugins/database/application/cell/bloc/date_cell_editor_bloc.dart (hunk 1 above)
Comment (richardshiue) [reply frontend/appflowy_flutter/lib/plugins/database/application/cell/bloc/date_cell_editor_bloc.dart:221]: Done

File: frontend/appflowy_flutter/lib/workspace/presentation/widgets/date_picker/appflowy_date_picker.dart (hunk 2 above)
Comment (richardshiue) [reply frontend/appflowy_flutter/lib/workspace/presentation/widgets/date_picker/appflowy_date_picker.dart:110]: Done

File: frontend/appflowy_flutter/lib/workspace/presentation/widgets/date_picker/mobile_appflowy_date_picker.dart (hunk 3 above)
Comment (richardshiue) [reply frontend/appflowy_flutter/lib/workspace/presentation/widgets/date_picker/mobile_appflowy_date_picker.dart:190]: For cubit/bloc, moving the logic to the cubit/bloc meant that triggering the callbacks required emitting extra states to have a BlocListener help do this. This resulted in a lot of boiler plate code in the widget and the cubit. And when the widget updates, that usually means the bloc is reinstantiated, or we need to update it separately

For an abstracted widget, it works a bit better, killing off ~120 lines. I think having an abstraction is overkill for this though and it makes the build methods hard to follow. What do you think of this? Or do you have something else in mind? https://github.com/AppFlowy-IO/AppFlowy/commit/9f682ca0d1a0b4e9af9662cc559f6d887597f289
